pkg/sql/copy/copy_test: TestLargeDynamicRows failed

pkg/sql/copy/copy_test.TestLargeDynamicRows [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9632591?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9632591?buildTab=artifacts#/) on release-23.1 @ [6cce6c746150307a9ecf3b529dfd633a6985c110](https://github.com/cockroachdb/cockroach/commits/6cce6c746150307a9ecf3b529dfd633a6985c110):


```
=== RUN   TestLargeDynamicRows
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/53491fdf16955d5896449a680358dad1/logTestLargeDynamicRows358695454
    test_log_scope.go:79: use -show-logs to present logs inline
    copy_test.go:594: 
        	Error Trace:	pkg/sql/copy/copy_test/pkg/sql/copy/copy_test.go:594
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestLargeDynamicRows
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/53491fdf16955d5896449a680358dad1/logTestLargeDynamicRows358695454
--- FAIL: TestLargeDynamicRows (1.81s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLargeDynamicRows.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27058

--- a/pkg/sql/copy/copy_test.go
+++ b/pkg/sql/copy/copy_test.go
@@ -548,6 +548,10 @@ func TestLargeDynamicRows(t *testing.T) {
 			batchNumber++
 			return nil
 		},
+		CopyFromInsertRetry: func() error {
+			batchNumber--
+			return nil
+		},
 	}
 	s, _, _ := serverutils.StartServer(t, params)
 	defer s.Stopper().Stop(ctx)

--- a/pkg/sql/copy_from.go
+++ b/pkg/sql/copy_from.go
@@ -642,9 +642,9 @@ func (c *copyMachine) processCopyData(ctx context.Context, data string, final bo
 		// If we have a full batch of rows or we have exceeded maxRowMem process
 		// them. Only set finalBatch to true if this is the last
 		// CopyData segment AND we have no more data in the buffer.
-		if len := c.currentBatchSize(); c.rowsMemAcc.Used() > c.maxRowMem || len == c.copyBatchRowSize || batchDone {
+		if len := c.currentBatchSize(); len > 0 && (c.rowsMemAcc.Used() > c.maxRowMem || len >= c.copyBatchRowSize || batchDone) {
 			if len != c.copyBatchRowSize {
-				log.VEventf(ctx, 2, "copy batch of %d rows flushing due to memory usage %d > %d", c.batch.Length(), c.rowsMemAcc.Used(), c.maxRowMem)
+				log.VEventf(ctx, 2, "copy batch of %d rows flushing due to memory usage %d > %d", len, c.rowsMemAcc.Used(), c.maxRowMem)
 			}
 			if err := c.processRows(ctx, final && c.buf.Len() == 0); err != nil {
 				return err
@@ -1026,6 +1026,11 @@ func (c *copyMachine) insertRows(ctx context.Context, finalBatch bool) error {
 			// NOTE: in theory we can also retry if c.insertRows == 0.
 			if c.implicitTxn && !c.p.SessionData().CopyFromAtomicEnabled && c.p.SessionData().CopyFromRetriesEnabled && errIsRetriable(err) {
 				log.SqlExec.Infof(ctx, "%s failed on attempt %d and is retrying, error %+v", c.copyFromAST.String(), r.CurrentAttempt(), err)
+				if c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry != nil {
+					if err := c.p.ExecCfg().TestingKnobs.CopyFromInsertRetry(); err != nil {
+						return err
+					}
+				}
 				continue
 			}
 			return err

--- a/pkg/sql/exec_util.go
+++ b/pkg/sql/exec_util.go
@@ -1613,6 +1613,9 @@ type ExecutorTestingKnobs struct {
 	// BeforeCopyFromInsert, if set, will be called during a COPY FROM insert statement.
 	BeforeCopyFromInsert func() error
 
+	// CopyFromInsertRetry, if set, will be called when a COPY FROM insert statement is retried.
+	CopyFromInsertRetry func() error
+
 	// ForceSQLLivenessSession will force the use of a sqlliveness session for
 	// transaction deadlines even in the system tenant.
 	ForceSQLLivenessSession bool

